### ISSUE TITLE
Sentry.Extensions.Logging Ignores Global Log Filters

### ISSUE THREAD

**Issue by mattico** [getsentry/sentry-dotnet#3847]
### Problem Statement

In my app I have global filters configured for certain libraries namespaces whose debug-level logs are spammy and not-actionable. The reason I configured them as global filters is because I want them to apply to all loggers, including Sentry's loggers. I spent quite some time confused about why Sentry was logging them anyway, before discovering [sentry ignores such filtering intentionally](https://github.com/getsentry/sentry-dotnet/pull/1992).

I disagree quite strongly with this choice of defaults. It makes sentry work differently from any other (Microsoft.Extensions.Logging) logger I'm aware of. It is also different from Sentry's Python logging integration, which honors the logging level. I understand that users who are unfamiliar with Microsoft.Extensions.Logging may be surprised that messages below Information are filtered by default. However the solution for that is a single line which can be included in documentation and sample programs.

Global filtering is a *useful feature* of Microsoft.Extensions.Logging, and it would continue to be useful with Sentry in addition to Sentry's internal filtering options.

### Solution Brainstorm

The most correct solution in my opinion would be to revert #1992 and explain in documentation and samples how users can configure the filtering to send all messages to Sentry if they choose.

However with an eye to backward compatibility, perhaps an option `BypassGlobalLogFilters` could be added to `SentryLoggingOptions` which defaults to `true`. Documentation should mention that sentry bypasses global filters by default.

**Comment by jamescrosswell** [comment]
Hey @mattico, thanks for the feedback.

From what I can tell, [this comment](https://github.com/getsentry/sentry-dotnet/issues/1986#issuecomment-1279625603) explains the original change:
> In a scenario where there are other loggers than just Sentry, we'd hopefully not need to set trace-level logging on by default in order for Sentry's logger to pick up trace logs as breadcrumbs. 

> The most correct solution in my opinion would be to revert https://github.com/getsentry/sentry-dotnet/pull/1992 and explain in documentation and samples how users can configure the filtering to send all messages to Sentry if they choose.

I'm wondering if there is any justification for Sentry to behave differently to other loggers and possibly there is.

If you're sending a bunch of logs to something like Splunk, you might configure a reasonably conservative log level in the interests of reducing costs. By contrast, the logs that Sentry collects are mostly used to create bread crumbs **_if_** there is an error... and when that happens, you probably want as much information as possible. So I could see how people might want different log levels for Sentry vs other log consumers. 

This could be done using [provider specific filters](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.filterloggingbuilderextensions.addfilter?view=net-9.0-pp#microsoft-extensions-logging-filterloggingbuilderextensions-addfilter-1(microsoft-extensions-logging-iloggingbuilder-system-func((microsoft-extensions-logging-loglevel-system-boolean))))... although the log levels will have to be configured in multiple places (since logs are not the only source of breadcrumbs in Sentry). 

It's hard to know all the different ways people are using Sentry and say, with certainty, that this would be better or worse for everyone. 

> However with an eye to backward compatibility, perhaps an option BypassGlobalLogFilters could be added to SentryLoggingOptions which defaults to true. Documentation should mention that sentry bypasses global filters by default.

That sounds like a safe option without any downsides that I can see. @bruno-garcia any thoughts?

@mattico to unblock you immediately, you can use the  [`SetBeforeBreadcrumb` callback on the Sentry options](https://docs.sentry.io/platforms/dotnet/enriching-events/breadcrumbs/#customize-breadcrumbs) to filter out breadcrumbs that you're not interested in.

**Comment by bruno-garcia** [comment]
The comments explain the thought process of the author:

```
  // All logs should flow to the SentryLogger, regardless of level.
  // Filtering of events is handled in SentryLogger, using SentryOptions.MinimumEventLevel
  // Filtering of breadcrumbs is handled in SentryLogger, using SentryOptions.MinimumBreadcrumbLevel
```

5.0.0 was a good target to change the default behavior. Not sure an option to bring back the old behavior is needed.

**Comment by Flash0ver** [comment]
We have revisited the `Sentry.Extensions.Logging` filtering behavior in `v6.0.0`.
However, we did double down on the existing behavior, which had some inconsistencies across integrations that are now made consistent.
The new [Structured Logs](https://docs.sentry.io/platforms/dotnet/guides/extensions-logging/logs/) feature from Sentry, however, does indeed tie into the Filtering of `Microsoft.Extensions.Logging`, to indeed be consistent within the ecosystem.

The justification being that our `Microsoft.Extensions.Logging` integration serves two scenarios:
- Reporting **Breadcrumbs** and **Events** out of Components through the interface of `Microsoft.Extensions.Logging`
  - these are configurable via (in namespace `Sentry.Extensions.Logging`)
    - `SentryLoggingOptions.MinimumBreadcrumbLevel`
    - `SentryLoggingOptions.MinimumEventLevel`
    - `SentryLoggingOptionsExtensions.AddLogEntryFilter(SentryLoggingOptions, ILogEntryFilter)`
    - `SentryLoggingOptionsExtensions.AddLogEntryFilter(SentryLoggingOptions, Func<string, LogLevel, EventId, Exception?, bool>)`
- Capturing trace-connected structured **Logs**
  - Configurable via existing `Microsoft.Extensions.Logging` idioms
    - `Microsoft.Extensions.Configuration`/`Microsoft.Extensions.Options`
      - e.g. `appsettings.{Environment}.json`
    - Filtering via `Microsoft.Extensions.Logging.FilterLoggingBuilderExtensions.AddFilter`

Configuring both features via the same mechanism would conflate two concerns,
and we'd loose a bit of fidelity in filtering.
Arguably, this is indeed more complex and potentially confusing.

**Comment by Flash0ver** [comment]
To be fair, we can be more explicit about both concerns/scenarios in our documentation. See #4952.

**Comment by mattico** [comment]
@Flash0ver yes, thank you for the thorough explainer! I actually upgraded my project to 6.0 last week and while it's not exactly what I wanted it does work much better for me.

**Comment by Flash0ver** [comment]
Glad to hear ... and thanks for your input ... it definitely helped us to make the system more internally consistent.

**Comment by Flash0ver** [comment]
The concerns are absolutely valid, though!
"Logs to Breadcrumbs and Events" come from a time where we didn't feature [Logs](https://sentry.io/product/logs/) yet.
But now, with **Logs** in the product, there is indeed a confusion between Breadcrumbs+Events and Structured Logs.

In the next major release - `v7` - we may simplify all our logging integrations, across all platforms and ecosystems.
We will definitely keep the idiomatic way of Structured Logging and its Configuration/Options,
but we may move, remove or rework Breadcrumbs and Events across all languages that Sentry supports.

**Comment by mattico** [comment]
@Flash0ver agreed.

I had to do some testing/investigation to understand how logs/breadcrumbs/events fit together and how I can use them. The conclusion I came to for my application was to configure logs "less-verbosely" compared to breadcrumbs. Since structured logs are always\* sent I want them to be relatively infrequent for performance and billing reasons, and to improve the signal-to-noise ratio on sentry.io. Since breadcrumbs are only sent in connection to events/errors (right?) I can afford to enable more verbose breadcrumbs to give more context to issues, where they're more likely to be relevant. My Sentry and Logging configuration has gotten a bit complicated to get this to where I want it. I'd welcome a redesign I think. I could elaborate on how my app is set up if you're interested.

The other thing I had to investigate, and which could be better documented IMO, is what causes issues to be created on Sentry.io or Errors in Spotlight. I had to do some trial and error testing to figure out which created issues:
- `LogError(String)`/`LogError(Exception)` (For a while I thought only logs with Exception objects created sentry issues but I think I was mistaken.)
- `LogWarning(String)`/`LogWarning(Exception)`
- `SentrySdk.CaptureMessage()` - Does the `SentryLevel` matter? (no)
- `SentrySdk.CaptureException(...)`
- `span.Finish(Exception)`/`transaction.Finish(Exception)`
- ...

**Comment by jamescrosswell** [comment]
> The other thing I had to investigate, and which could be better documented IMO, is what causes issues to be created on Sentry.io

@mattico that's determined by the [MinimumEventLevel](https://docs.sentry.io/platforms/dotnet/guides/extensions-logging/#minimumeventlevel)

It looks like that option is covered in the [Quick Start guide](https://docs.sentry.io/platforms/dotnet/guides/extensions-logging/#minimumeventlevel) but isn't mentioned anywhere in the [Extended Configuration | Options](https://docs.sentry.io/platforms/dotnet/guides/extensions-logging/configuration/options/). @Flash0ver we could look at adding integration specific properties to the Extended Configuration | Options (not just for the logging integrations but for all integrations).
